testsuite/tests/callback/test_finaliser_gc.ml

```diff
@@ -0,0 +1,31 @@
+(* TEST
+*)
+
+let rec tak (x, y, z as _tuple) =
+  if x > y then tak(tak (x-1, y, z), tak (y-1, z, x), tak (z-1, x, y))
+           else z
+
+let finaliser_status = ref "Finaliser NOT run"
+let z = ref (0, 1, 2, 3, 4, 5, 6, 7)
+
+let force_gc_fn _ =
+  finaliser_status := "Finaliser has run!";
+  Gc.full_major ()
+
+let trigger_finaliser () =
+  (* Construct finaliser which when run will force
+     a major cycle *)
+  Gc.finalise force_gc_fn (ref 0);
+  (* Allocate a block in the minor heap *)
+  let s = String.make 5 'b' in
+  (* Spin on the minor heap allocating but keep [s] in a
+    register and force a major cycle such that the
+    finaliser runs *)
+  for x = 0 to 10_000_000 do
```

Comment:
Presumably this is dependent on the size of the minor heap. If I set a minor heap a bit larger than 640mb I think this will fail as well?

Maybe we should use the current Gc minor heap size to set the iteration count.

Reply:
I've made this less dependent on the minor heap size and also it will complete a bit quicker when things work as expected. I flag when the finaliser has run to break the loop. The upper bound is there to prevent the test hanging in the (hopefully unlikely) case that the finaliser code gets broken. 

The upper bound will prevent you having arbitrarily large minor heaps; we are allocating of order 8Bn words if this doesn't force the finaliser to run then the test will not work as expected. 

I've verified that the changes keep things in the register and correctly fail on the old buggy code before #530.